batcher/aligned-batcher/src/types/batch_state.rs

```diff
@@ -0,0 +1,159 @@
+    }
+
+    /// Checks if the entry is valid
+    /// An entry is valid if there is no entry with the same sender, lower nonce and a lower fee
+    pub(crate) fn replacement_entry_is_valid(
+        &mut self,
+        replacement_entry: &BatchQueueEntry,
+    ) -> bool {
+        let replacement_max_fee = replacement_entry.nonced_verification_data.max_fee;
+        let nonce = replacement_entry.nonced_verification_data.nonce;
+        let sender = replacement_entry.sender;
+
+        debug!(
+            "Checking validity of entry with sender: {:?}, nonce: {:?}, max_fee: {:?}",
+            sender, nonce, replacement_max_fee
+        );
+
+        // it is a valid entry only if there is no entry with the same sender, lower nonce and a lower fee
+        !self.batch_queue.iter().any(|(entry, _)| {
+            entry.sender == sender
+                && entry.nonced_verification_data.nonce < nonce
+                && entry.nonced_verification_data.max_fee < replacement_max_fee
+        })
+    }
+
+    pub(crate) fn get_user_min_fee_in_batch(&self, addr: &Address) -> U256 {
+        self.batch_queue
+            .iter()
+            .filter(|(e, _)| &e.sender == addr)
+            .map(|(e, _)| e.nonced_verification_data.max_fee)
+            .min()
+            .unwrap_or(U256::max_value())
+    }
+
+    pub(crate) fn update_user_min_fee(
+        &mut self,
+        addr: &Address,
+        new_min_fee: U256,
+    ) -> Option<U256> {
+        if let Entry::Occupied(mut user_state) = self.user_states.entry(*addr) {
+            user_state.get_mut().min_fee = new_min_fee;
+            return Some(new_min_fee);
+        }
+        None
+    }
+
+    pub(crate) fn update_user_proof_count(
+        &mut self,
+        addr: &Address,
+        new_proof_count: usize,
+    ) -> Option<usize> {
+        if let Entry::Occupied(mut user_state) = self.user_states.entry(*addr) {
+            user_state.get_mut().proofs_in_batch = new_proof_count;
+            return Some(new_proof_count);
+        }
+        None
+    }
+
+    /// Updates the user with address `addr` with the provided values of `new_nonce`, `new_min_fee` and
+    /// `new_proof_count`.
+    /// If state is updated successfully, returns the updated values inside a `Some()`
+    /// If the address was not found in the user states, returns `None`
+    pub(crate) fn update_user_state(
+        &mut self,
+        addr: &Address,
+        new_nonce: U256,
+        new_min_fee: U256,
+        new_proof_count: usize,
+    ) -> Option<(U256, U256, usize)> {
+        let updated_nonce = self.update_user_nonce(addr, new_nonce);
+        let updated_min_fee = self.update_user_min_fee(addr, new_min_fee);
+        let updated_proof_count = self.update_user_proof_count(addr, new_proof_count);
+
+        match (updated_nonce, updated_min_fee, updated_proof_count) {
+            (Some(_), Some(_), Some(_)) => Some((new_nonce, new_min_fee, new_proof_count)),
+            _ => None,
+        }
+    }
```

Comment:
Same as above, if you're not using the inner value maybe you can check those variables with `is_some()`.